### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by glenngijsberts**
# Description

This PR will introduce Typescript!

## How to contribute

* Checkout PR
* Make sure you are on the latest commit
* Do not force push so we can all safely use `git pull`

**Component**

* Change component to `.tsx`
* Make sure to type the file. Use `enum` types as much as possible
* Export Typescript related stuff before declaration like: `export interface AlertProps`
* Export components at the bottom of the file like: `export { Accordion, AccordionPanel ... }`

**Storybook**

* Change storybook file to `.tsx` and fix potential typings such as applying enums
* Export stories at the bottom as much as possible

**Tests** 
* Change test file to `.tsx` if it contains jsx code, otherwise change it to `.ts`
* Fix the tests if needed, such as applying enums

For inspiration and/or examples you can take a look at components that are already done such as `Alert` & `Accordion`. The goal is refactoring the code to TypeScript, without making changes to the functionality. However, if you may need to remove some code or do need to change functionality you should mark this as a breaking change in your commit, using `yarn commit`.

## Changes

- [x] Enable typescript
- [x] Generate declarations
- [x] Enable Typescript for Jest tests 
- [x] Hooks - Typescript ready
- [x] Utils - Typescript ready
- [x] Use TSDX

- [x] Fix strict warning and errors

- [x] Components:
- - [x] Alert
- - [x] Avatar
- - [x] Accordion
- - [x] Banner
- - [x] Text
- - [x] BaseStyles
- - [x] StarRating
- - [x] Button
- - [x] Card
- - [x] ContentDialog
- - [x] Panel
- - [x] Pill
- - [x] BaseButton
- - [x] Collapsible
- - [x] Toggle
- - [x] Checkbox
- - [x] Select
- - [x] DateInput
- - [x] MoneyInput
- - [x] PhoneInput
- - [x] TimeInput
- - [x] Combobox
- - [x] ContentTransition
- - [x] Skeleton
- - [x] Toast
- - [x] Cover
- - [x] CustomDropdown
- - [x] Dialog
- - [x] OnboardingDialog
- - [x] Flag
- - [x] global-styles
- - [x] Heading
- - [x] HorizontalScroll
- - [x] Image
- - [x] Input
- - [x] Logo
- - [x] MenuButton
- - [x] Portal
- - [x] SpeechBubble
- - [x] Spinner
- - [x] Tabs 
- - [x] VisuallyHidden
- - [x] Tooltip

## Breaking changes

* `easingFunctions` are removed from `theme.js`
* `debounce` is removed
* The `handler` parameter of `useOnClickOutside` is no longer exposing the `event` after firing

### CODE REVIEW

File: src/components/Pill/index.tsx [hunk 1]
```diff
@@ -46,29 +46,18 @@ const LeftAdornment = styled.span`
...
-  return (
-    <Container className="pill" {...props}>
-      {props.leftAdornment && (
-        <LeftAdornment>{props.leftAdornment}</LeftAdornment>
-      )}
-      {children}
-    </Container>
-  )
+interface PillProps {
+  variant: String
```
Comment (RobVermeer) [inline src/components/Pill/index.tsx]: I think you can do the following, so it's an enum list of variants.

```typescript
enum Variant {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  danger = 'danger',
  warning = 'warning',
  disabled = 'disabled',
}

interface PillProps {
  variant: Variant,
  leftAdornment: Element,
}
```

**Comment by rppld** [comment]
Check out [TSDX](https://github.com/formium/tsdx). Might simplify the bundling setup quite a bit.

File: src/components/Alert/index.tsx [hunk 2]
```diff
@@ -19,23 +18,41 @@ import {
...
+  label: string
+  onClick: (event: React.MouseEvent<HTMLElement>) => void
+}
+
+export enum AlertVariant {
+  info = 'info',
+  success = 'success',
+  error = 'error',
+  warning = 'warning',
+}
```
Comment (rppld) [inline src/components/Alert/index.tsx:31]: There’s a [case against](https://blog.logrocket.com/why-typescript-enums-suck/) using enums in TypeScript. Might be clearer and less error-prone to use string union types.
```suggestion
export type AlertVariant = 'info' | 'succes' | 'error' | 'warning'
```

Comment (glenngijsberts) [reply src/components/Alert/index.tsx:31]: Thanks Philipp! Let's discuss tomorrow @RobVermeer @Roohn ;)

**Review by Roohn — APPROVED** [review]

**Pull request merged**